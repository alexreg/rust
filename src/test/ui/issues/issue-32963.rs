use std::mem;

trait Misc {}

fn size_of_copy<T: Copy + ?Sized>() -> usize { mem::size_of::<T>() }

fn main() {
    size_of_copy::<dyn Misc + Copy>();
    //~^ ERROR only auto traits can be used as additional traits in a trait object [E0225]
    //~| ERROR the trait bound `dyn std::marker::Copy: std::marker::Copy` is not satisfied [E0277]
    //~| ERROR the trait `std::marker::Copy` cannot be made into an object [E0038]
}
