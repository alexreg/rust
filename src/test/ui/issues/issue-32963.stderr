error[E0225]: only auto traits can be used as additional traits in a trait object
  --> $DIR/issue-32963.rs:8:31
   |
LL |     size_of_copy::<dyn Misc + Copy>();
   |                        ----   ^^^^
   |                        |      |
   |                        |      additional non-auto trait
   |                        |      trait alias used in trait object type (additional use)
   |                        first non-auto trait
   |                        trait alias used in trait object type (first use)

error[E0277]: the trait bound `dyn std::marker::Copy: std::marker::Copy` is not satisfied
  --> $DIR/issue-32963.rs:8:5
   |
LL | fn size_of_copy<T: Copy + ?Sized>() -> usize { mem::size_of::<T>() }
   |    ------------    ---- required by this bound in `size_of_copy`
...
LL |     size_of_copy::<dyn Misc + Copy>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::marker::Copy` is not implemented for `dyn std::marker::Copy`

error[E0038]: the trait `std::marker::Copy` cannot be made into an object
  --> $DIR/issue-32963.rs:8:20
   |
LL |     size_of_copy::<dyn Misc + Copy>();
   |                    ^^^^^^^^^^^^^^^ the trait `std::marker::Copy` cannot be made into an object
   |
   = note: the trait cannot require that `Self : Sized`

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0038, E0225, E0277.
For more information about an error, try `rustc --explain E0038`.
